src/error.rs

```diff
@@ -97,29 +97,166 @@ impl From<OddLengthStringError> for ToBytesError {
+                    let (hi, lo) = crate::iter::byte_to_hex_digits(byte);
+                    if !self.is_hi {
+                        let vals = [self.lo];
+                        error(&vals, self.pos + 1).or_else(|| {
+                            let vals = [self.lo, hi];
+                            error(&vals, self.pos + 1).or_else(|| {
+                                let vals = [self.lo, hi, lo];
+                                error(&vals, self.pos + 1)
+                            })
+                        })
+                    } else {
+                        let vals = [self.hi];
+                        error(&vals, self.pos).or_else(|| {
+                            let vals = [self.hi, self.lo];
+                            error(&vals, self.pos).or_else(|| {
+                                let vals = [self.hi, self.lo, hi];
+                                error(&vals, self.pos).or_else(|| {
+                                    let vals = [self.hi, self.lo, hi, lo];
+                                    error(&vals, self.pos)
+                                })
+                            })
+                        })
+                    }
+                }
+                // The next two bytes happen to be invalid ASCII.
+                Err(e) =>
+                    if !self.is_hi {
+                        let vals = [self.lo, e.hi];
+                        error(&vals, self.pos + 1).or_else(|| {
+                            let vals = [self.lo, e.hi, e.lo];
+                            error(&vals, self.pos + 1)
+                        })
+                    } else {
+                        let vals = [self.hi, self.lo, e.hi];
+                        error(&vals, self.pos).or_else(|| {
+                            let vals = [self.hi, self.lo];
+                            error(&vals, self.pos).or_else(|| {
+                                let vals = [self.hi, self.lo, e.hi];
+                                error(&vals, self.pos).or_else(|| {
+                                    let vals = [self.hi, self.lo, e.hi, e.lo];
+                                    error(&vals, self.pos)
+                                })
+                            })
+                        })
+                    },
+            },
+            // Invalid character was for the last character in the input string.
+            None =>
+                if !self.is_hi {
+                    let vals = [self.lo];
+                    error(&vals, self.pos + 1)
+                } else {
+                    let vals = [self.hi];
+                    error(&vals, self.pos).or_else(|| {
+                        let vals = [self.hi, self.lo];
+                        error(&vals, self.pos)
+                    })
+                },
+        };
+        rich_error.unwrap_or(InvalidCharError {
+            invalid: if self.is_hi { self.hi.into() } else { self.lo.into() },
+            pos: self.pos,
+        })
```

Comment:
In ce28e9ba01cda818439b36108e072435715552fc:

This logic is really confusing. It seems like if the original "bad digit" was an ASCII byte then that byte can just be promoted to `char` and we're done. If not, we need to read more. So you could have an initial `if self.invalid_byte() < 128 { return InvalidCharError { invalid: self.invalid_byte().into(), pos: self.pos() }`. (And ofc `self.invalid_byte` should be renamed, see above.) So you can do an early return in this case and save yourself at least one level of nesting. (And also fix the bug here `self.pos` is wrong and you actually meant to return `self.pos()`.)